File: locate/android/app/src/main/java/com/example/locate/domain/model/ServerMessage.kt
package com.example.locate.domain.model

import org.json.JSONObject

/**
 * 服务器 WebSocket 消息解析
 */
sealed class ServerMessage {
    data class LoginSuccess(
        val token: String,
        val nickname: String
    ) : ServerMessage()

    data class UserList(val users: List<User>) : ServerMessage()

    data class UserJoined(val username: String) : ServerMessage()

    data class UserLeft(val username: String) : ServerMessage()

    data class TargetUpdate(
        val username: String,
        val targetLat: Double?,
        val targetLng: Double?
    ) : ServerMessage()

    data class PositionUpdate(
        val username: String,
        val lat: Double,
        val lng: Double,
        val heading: Float = 0f
    ) : ServerMessage()

    data class Error(val message: String) : ServerMessage()

    companion object {
        fun parse(json: String): ServerMessage? {
            return try {
                val obj = JSONObject(json)
                when (obj.getString("type")) {
                    "login_success" -> LoginSuccess(
                        token = obj.getString("token"),
                        nickname = obj.optString("nickname", "")
                    )
                    "user_list" -> {
                        val users = mutableListOf<User>()
                        val arr = obj.optJSONArray("users") ?: return UserList(emptyList())
                        for (i in 0 until arr.length()) {
                            val u = arr.getJSONObject(i)
                            users.add(User(
                                username = u.getString("username"),
                                lat = u.getDouble("lat"),
                                lng = u.getDouble("lng"),
                                heading = u.optDouble("heading", 0.0).toFloat(),
                                nickname = u.optString("nickname", "").takeIf { it.isNotEmpty() },
                                targetLat = u.optDouble("target_lat").takeIf { !u.isNull("target_lat") },
                                targetLng = u.optDouble("target_lng").takeIf { !u.isNull("target_lng") }
                            ))
                        }
                        UserList(users)
                    }
                    "user_joined" -> UserJoined(obj.getString("username"))
                    "user_left" -> UserLeft(obj.getString("username"))
                    "update_target" -> TargetUpdate(
                        username = obj.getString("username"),
                        targetLat = obj.optDouble("target_lat").takeIf { !obj.isNull("target_lat") },
                        targetLng = obj.optDouble("target_lng").takeIf { !obj.isNull("target_lng") }
                    )
                    "update_position" -> PositionUpdate(
                        username = obj.getString("username"),
                        lat = obj.getDouble("lat"),
                        lng = obj.getDouble("lng"),
                        heading = obj.optDouble("heading", 0.0).toFloat()
                    )
                    "error" -> Error(obj.optString("message", "未知错误"))
                    else -> null
                }
            } catch (e: Exception) {
                null
            }
        }
    }
}

File: locate/ios/locate/models/ServerMessage.swift
import Foundation

/// 服务器 WebSocket 消息解析
/// 和 Android ServerMessage.kt 完全一致的解析逻辑
enum ServerMessage {
    case loginSuccess(token: String, nickname: String)
    case userList(users: [User])
    case userJoined(username: String, lat: Double, lng: Double, heading: Float, nickname: String?, targetLat: Double?, targetLng: Double?)
    case userLeft(username: String)
    case targetUpdate(username: String, targetLat: Double?, targetLng: Double?)
    case positionUpdate(username: String, lat: Double, lng: Double, heading: Float)
    case error(message: String)

    static func parse(_ jsonString: String) -> ServerMessage? {
        guard let data = jsonString.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = obj["type"] as? String else {
            return nil
        }

        switch type {
        case "login_success":
            let token = obj["token"] as? String ?? ""
            let nickname = obj["nickname"] as? String ?? ""
            return .loginSuccess(token: token, nickname: nickname)

        case "user_list":
            guard let usersArray = obj["users"] as? [[String: Any]] else {
                return .userList(users: [])
            }
            let users = usersArray.compactMap { parseUser($0) }
            return .userList(users: users)

        case "user_joined":
            return .userJoined(
                username: obj["username"] as? String ?? "",
                lat: obj["lat"] as? Double ?? 0,
                lng: obj["lng"] as? Double ?? 0,
                heading: Float(obj["heading"] as? Double ?? 0),
                nickname: obj["nickname"] as? String,
                targetLat: obj["target_lat"] as? Double,
                targetLng: obj["target_lng"] as? Double
            )

        case "user_left":
            return .userLeft(username: obj["username"] as? String ?? "")

        case "update_target":
            return .targetUpdate(
                username: obj["username"] as? String ?? "",
                targetLat: obj["target_lat"] as? Double,
                targetLng: obj["target_lng"] as? Double
            )

        case "update_position":
            return .positionUpdate(
                username: obj["username"] as? String ?? "",
                lat: obj["lat"] as? Double ?? 0,
                lng: obj["lng"] as? Double ?? 0,
                heading: Float(obj["heading"] as? Double ?? 0)
            )

        case "error":
            return .error(message: obj["message"] as? String ?? "未知错误")

        default:
            return nil
        }
    }

    private static func parseUser(_ dict: [String: Any]) -> User? {
        guard let username = dict["username"] as? String else { return nil }
        return User(
            username: username,
            lat: dict["lat"] as? Double ?? 0,
            lng: dict["lng"] as? Double ?? 0,
            heading: dict["heading"] as? Float ?? Float(dict["heading"] as? Double ?? 0),
            nickname: dict["nickname"] as? String,
            targetLat: dict["target_lat"] as? Double,
            targetLng: dict["target_lng"] as? Double
        )
    }
}